src/v/transform/api.cc

```diff
@@ -698,14 +698,24 @@ ss::future<std::error_code> service::deploy_transform(
     auto [key, offset] = result.value();
     meta.uuid = key;
     meta.source_ptr = offset;
-    meta.offset_options = model::transform_offset_options{
-      // Set the transform to start processing new records starting now,
-      // this is the default expectations for developers, as once deploy
-      // completes, they should be able to produce without waiting for the
-      // vm to start. If we start from the end of the log, then records produced
-      // between now and the vm start would be skipped.
-      .position = model::new_timestamp(),
-    };
+
+    // Use latest_offset as a sentinel value during user-driven deploy. We won't
+    // expose this option through the API anyway, and already-serialized
+    // transform metadata (i.e. legacy deployments) won't traverse this code.
+    // Otherwise, respect whatever offset was included in the request.
+    if (
+      std::get_if<model::transform_offset_options::latest_offset>(
+        &meta.offset_options.position)
+      != nullptr) {
+        meta.offset_options = model::transform_offset_options{
+          // Set the transform to start processing new records starting now,
+          // this is the default expectations for developers, as once deploy
+          // completes, they should be able to produce without waiting for the
+          // vm to start. If we start from the end of the log, then records
+          // produced between now and the vm start would be skipped.
```

Comment:
I'm wondering if by the same logic, we should be resolving the offset delta to a `model::offset` here instead of doing it in `transform::processor::load_latest_committed()`. For example, offset from the end `-0` is equivalent to `latest` yet they behave slightly differently. But maybe it's not a big deal since the processor is started "soon enough"?

Would it make sense to move the offset delta resolution and latest resolution up to the admin API handler and change the `transform_offset_options` to hold a fixed position `serde::variant<model::timestamp, model::offset> position`?

Reply:
The trouble is that offsets are committed per transform + partition, so we actually need to compute a concrete offset when each vm spins up. So I think `load_latest_committed` is the right place for the calculation - "give me the last committed offset on this partition, and if none exists calculate one based on my config".

To the similarity between `latest` and `-0`, it's true, but `latest` isn't exposed at the API layer. As I understand it, it was meant primarily as a compatibility default for deployed transforms that predate the options field. So "-oo from end" is inclusive of that behavior but intended as a user-facing knob.

Reply:
Ah, I didn't think this through carefully. `-0` will _not_ give the same result as `latest`. `offset{"format": "from_end", "value": 0}` winds up as a named int64, so we will eventually interpret it as plain old `0` and start processing from the beginning of the partition.

We should cover for this somehow - probably just by rejecting `--from-offset -0` in rpk. I don't think it's ever particularly useful compared to the default behavior. i.e. "commit offset at deploy time" vs "commit offset at VM start time".

Reply:
Thanks for the explanation. I like the more detailed variant type you went with 👍